data/wikipron/src/scrape.py

```diff
@@ -18,46 +18,127 @@
 from codes import LANGUAGES_PATH, LOGGING_PATH
 
 
+def _whitelist_reader(path: str) -> Iterator[str]:
+    # Read-in whitelist file.
+    with open(path, "r") as source:
+        for line in source:
+            line = re.sub(r"\s*#.*$", "", line)  # Removes comments from line.
+            yield line.rstrip()
+
+
+def _filter(word: str, pron: str, phones: FrozenSet[str]) -> bool:
+    # Determines if gloss is valid given phone set.
+    these_phones = frozenset(pron.split())
+    bad_phones = these_phones - phones
+    if bad_phones:
+        for phone in bad_phones:
+            logging.warning("Bad phone:\t%s\t(%s)", phone, word)
+        return False
+    else:
+        return True
+
+
 def _call_scrape(
-    lang_settings: Dict[str, str], config: wikipron.Config, tsv_path: str
+    lang_settings: Dict[str, str],
+    config: wikipron.Config,
+    tsv_path: str,
+    whitelist_set: FrozenSet[str] = None,
+    tsv_filtered_path: str = "",
 ) -> None:
     for unused_retries in range(10):
         with open(tsv_path, "w") as source:
             try:
-                for (word, pron) in wikipron.scrape(config):
-                    print(f"{word}\t{pron}", file=source)
+                scrape_results = wikipron.scrape(config)
+                # Given whitelist, opens up a second tsv for scraping.
+                if whitelist_set:
+                    with open(tsv_filtered_path, "w") as source_filtered:
+                        for (word, pron) in scrape_results:
+                            line = f"{word}\t{pron}"
+                            if _filter(word, pron, whitelist_set):
+                                print(line, file=source_filtered)
+                            print(line, file=source)
+                else:
+                    for (word, pron) in scrape_results:
+                        print(f"{word}\t{pron}", file=source)
                 return
             except (
                 requests.exceptions.Timeout,
                 requests.exceptions.ConnectionError,
             ):
                 logging.info(
-                    'Exception detected while scraping: "%s", "%s".',
+                    'Exception detected while scraping: "%s", "%s", "%s".',
                     lang_settings["key"],
                     tsv_path,
+                    tsv_filtered_path,
                 )
                 # Pauses execution for 10 min.
                 time.sleep(600)
-    # Log and remove TSVs for languages that failed
+    # Log and remove TSVs for languages that failed.
     # to be scraped within 10 retries.
     logging.info(
         'Failed to scrape "%s" within 10 retries. %s',
         lang_settings["key"],
         lang_settings,
     )
+    # Checks if second TSV was opened.
+    if os.path.exists(tsv_filtered_path):
```

Comment:
Suggest using try/except here instead of testing for existence. In theory it can help you avoid a weird race condition.

Reply:
Could you go into more detail? I didn't know that could happen here.

Reply:
There's no particular reason to think it would occur here but the file could *in theory* be materialized between line 84 and line 85, in which case it would not be removed, or it could be deleted between lines 84 and 85, in which case this would throw an uncaught exception. So the general recommendation is try and fail.